core/src/solution_submission/retry.rs

```diff
@@ -99,9 +106,44 @@ fn gas_price(estimated_price: U256, price_increase_count: u32, cap: U256) -> U25
     cap.min(U256::from(new_price as u128))
 }
 
+/// The output type of futures used in `select_all`.
+enum SelectFutureOutput {
+    SolutionSubmission(Result<(), MethodError>),
+    Timeout,
+}
+
+// Return the first future result, skipping nonce errors unless the nonce error is the last result.
+//
+// `futures` contains all solution submission attempts and optionally a timeout. Unfinished futures
+//  stay in the vector when the function returns.
+async fn first_result_ignoring_nonce_errors(
+    futures: &mut Vec<impl Future<Output = SelectFutureOutput> + Unpin>,
```

Comment:
We have to give the unfinished futures back to the caller. This can either be done with the mut reference or as you say by returning them back. I'll try it to see how it looks.

Reply:
Ah, I thought the caller was ignoring the unfinished futures. In that case I'm fine with how it is.